forge/forge/compiled_graph_state.py

```diff
@@ -166,35 +170,54 @@ class ProgramId(IntEnum):
 
 class CompiledModel:
     """
-    Callable object for running inference on the compiled model.
+    Callable object for running the compiled model on the device(s).
+
+    If the model is compiled for inference, only forward pass can be executed.
+    In case of training - forward, backward, loss and optimizer steps can be executed - depending on which of these
+    is compiled for the device, and which are set up to be ran separately on the CPU.
     """
 
     fwd_compiled_graph_state: CompiledGraphState
     bwd_compiled_graph_state: Optional[CompiledGraphState]
+
+    # Compiled flatbuffer binary composed of programs which execute compiled graphs (e.g., forward, backward, etc.)
     compiled_binary: Binary
+
     inputs: List[torch.Tensor]
+    outputs: Dict[str, torch.Tensor]
     intermediates: List[torch.Tensor]
+
+    # Original user-defined module.
     framework_module: AnyModule
-    loss_module: Optional[Module]
-    optimizer: Optional[torch.optim.Optimizer]
+
+    # Gradients coming from the loss.backward() - only used when loss is computed on CPU.
+    loss_grad: List[Optional[torch.Tensor]]
 
     def __init__(
         self,
         fwd_compiled_graph_state: CompiledGraphState,
-        bwd_compiled_graph_state: CompiledGraphState,
+        bwd_compiled_graph_state: Optional[CompiledGraphState],
         compiled_binary: Binary,
         framework_module: AnyModule,
-        loss_module: Optional[Module] = None,
-        optimizer: Optional[torch.optim.Optimizer] = None,
     ):
         self.fwd_compiled_graph_state = fwd_compiled_graph_state
         self.bwd_compiled_graph_state = bwd_compiled_graph_state
         self.compiled_binary = compiled_binary
         self.inputs = []
         self.framework_module = framework_module
-        self.loss_module = loss_module
-        self.optimizer = optimizer
         self.intermediates = []
+        self.loss_grad = [None for _ in range(len(fwd_compiled_graph_state.ordered_external_output_names))]
```

Comment:
very nit: maybe self.loss_grad = [None] * len(fwd...), or [None for _ in fwd....]?

Reply:
done